### PR TITLE
8335662: [AArch64] C1: guarantee(val < (1ULL << nbits)) failed: Field too big for insn

### PR DESCRIPTION
**Pull request by chadrako**
[JDK-8335662](https://bugs.openjdk.org/browse/JDK-8335662)

Crash occurs in C1 during OSR when copying locks from interpreter frame to compiled frame. All loads used immediate offset regardless of offset size causing crash when it is over the max size for the instruction (32760). Fix is to check the size before preforming the load and storing the offset in a register if needed.

I believe the risk is low because there will be no change to the instruction if the immediate offset fits in the load instruction. The instruction is only updated when the `offset_ok_for_immed` check fails which would cause the crash anyways

Confirmed that added test fails before patch and passes after

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335662](https://bugs.openjdk.org/browse/JDK-8335662): [AArch64] C1: guarantee(val &lt; (1ULL &lt;&lt; nbits)) failed: Field too big for insn (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21473/head:pull/21473` \
`$ git checkout pull/21473`

Update a local copy of the PR: \
`$ git checkout pull/21473` \
`$ git pull https://git.openjdk.org/jdk.git pull/21473/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21473`

View PR using the GUI difftool: \
`$ git pr show -t 21473`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21473.diff">https://git.openjdk.org/jdk/pull/21473.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21473#issuecomment-2411905116)

### CODE REVIEW

**Comment by theRealAph** [comment]
Thanks. Fix looks reasonable, but i think we need a regression test.

**Comment by chadrako** [comment]
Yeah working on adding a regression test

**Review by dean-long — CHANGES_REQUESTED** [review]
Why is part of the test a binary .class file?

File: test/hotspot/jtreg/compiler/c1/Test8335662.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8335662
+ * @summary Execute main() method
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c1/Test8335662.java:27]: Please use a more descriptive summary of the test.

File: test/hotspot/jtreg/compiler/c1/Test8335662.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @run main/othervm -XX:+UseOnStackReplacement -XX:TieredStopAtLevel=1
+ *      Test8335662
+ */
+
+import java.lang.reflect.Method;
+
+public class Test8335662 {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c1/Test8335662.java:35]: We don't use bug numbers for test names (anymore).

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
The class file is from the original bug report, it should be converted to a jasm file.

**Comment by theRealAph** [comment]
One thing for you to think about ifm you are interested in sone further work in this area..

This is a generic problem. It might be very beneficial to look for every base + immediate offset instruction, see if there is a possibility that there may be an overflow, and insert a `form_address()`.

File: test/hotspot/jtreg/compiler/c1/ComplexLockingAndMultiThreading.jasm [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  private static final Field lock:"Ljava/lang/Object;";
+
+  public Method "<init>":"()V"
+    stack  1 locals  1
+  {
+         aload_0;
+         invokespecial     Method java/lang/Object."<init>":"()V";
+         return;
+  }
+  public static Method main:"([Ljava/lang/String;)V"
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c1/ComplexLockingAndMultiThreading.jasm:13]: There's no need to have the main method in jasm, right? I think it's only `synchronizedMethod` which triggers the issue. So ideally, the jasm file is as compact as possible to ease future maintenance of the test.

Comment (eastig) [reply test/hotspot/jtreg/compiler/c1/ComplexLockingAndMultiThreading.jasm:13]: IMO we even don't need to run the code to trigger compilation. We can use WhiteBox to compile it.

**Review by TobiHartmann — APPROVED** [review]
Thanks for simplifying the test, very nice! The changes look good to me.

**Review by eastig — APPROVED** [review]
lgtm

**Comment by eastig** [comment]
> One thing for you to think about if you are interested in some further work in this area..
> 
> This is a generic problem. It might be very beneficial to look for every base + immediate offset instruction, see if there is a possibility that there may be an overflow, and insert a `form_address()`.

Hi @theRealAph,

https://bugs.openjdk.org/browse/JDK-8342736

We found there are around ~400 ldr calls and ~180 str calls that would need to be manually updated.
I found that we use offsets in C++ classes a lot:
```
$ cpu/aarch64 % grep _offset\(\) *.* | grep ldr | wc -l
250
$ cpu/aarch64 % grep _offset\(\) *.* | grep str | wc -l
94
$ cpu/aarch64 % grep _offset\(\) *.* | grep lea | wc -l
27
```

IMO, we can use `static_assert` for them. The problem is that the macro `offset_of` is not `constexpr`. Making it `constexpr` is not simple. The standard macro `offsetof` requires classes to have the standard layout. Most of our classes don't have the standard layout. We'll get warnings about this. As you wrote in the macro comments, we can disable warnings. 
I think can have something like this:
```
ldr(dst, Address(rmethod, create_mem_op_imm(Method, const_offset)));

#define create_mem_op_imm(klass, field_offset_func) \
([]() { \
 constexpr max_possible_offset = sizeof(klass);
 static_assert(Address::offset_ok_for_immed(max_possible_offset, 0)); \
 return klass::field_offset_func(); \
}())
```

If the size of a class fits into a memory instructions then any offset in it will fit. Class sizes greater than 32760 look insane to me.

**Pull request closed without merging**